File: doubly_circularly_linked_list.java
// package whatever; // don't place package name!

import java.io.*;
import java.util.ArrayList;

class DoublyCircNode{
  DoublyCircNode next;
  DoublyCircNode prev;
  int value;
  
  public DoublyCircNode(int value){
    this.value = value;
    this.next = null;
    this.prev= null;
  }
  
}

class DoublyCircLinkedList{
  DoublyCircNode last;
  public void append(int value){
    DoublyCircNode newNode = new DoublyCircNode(value);
    if(last == null){
      last = newNode;
      last.next = newNode;
      last.prev = newNode;
      return;
    }
    newNode.prev = last;
    newNode.next = last.next;
    last.next.prev = newNode;
    last.next = newNode;
    last = newNode;
  }
  
  public void prepend(int value){
    DoublyCircNode newNode = new DoublyCircNode(value);
    if(last == null){
      append(value);
      return;
    }
    newNode.prev = last;
    newNode.next = last.next;
    last.next.prev = newNode;
    last.next = newNode;
    return;
  }
  
  public int getLength(){
    if(last != null){
      DoublyCircNode current = last;
      int i = 1;
      while(current.next != last){
        i++;
        current = current.next;
      }
      return i;
    }else{
      return 0;
    }
  }
  
  public void insert(int value, int index){
    if(index == 0){
      prepend(value);
      return;
    }else if(index > getLength() || last == null){
      append(value);
      return;
    }
    else{
      DoublyCircNode current = last;
      DoublyCircNode newNode = new DoublyCircNode(value);
      int i = 1;
      while(current.next != last && i < index){
        current = current.next;
        i++;
      }
      if(current.next == last){
        newNode.prev = current;
        newNode.next = current.next;
        current.next.prev = newNode;
        current.next = newNode;
      }else{
        newNode.prev = current;
        newNode.next = current.next;
        current.next.prev = newNode;
        current.next = newNode;
      }
    }
  }
  public void deleteFirst(int value){
    if(last == null){
      return;
    }else if(last.next == last && last.value == value){
      last = null;
      return;
    }
    DoublyCircNode current = last;
    while(current.next != last && current.next.value != value){
      current = current.next;
    }
    if(current.next == last && last.value == value){
      last.prev.next = last.next;
      last.next.prev = last.prev;
      last = last.prev;
      return;
    }else if(current.next == last && !(last.value == value)){
      return;
    }else{
    current.next = current.next.next;
    current.next.prev = current;
    return;
    }
  }
  
  public void deleteAll(int value){
    if(last == null){
      return;
    }
    DoublyCircNode current = last.next;
    while(current != last){
      if(current.value == value){
        deleteFirst(value);
      }
      current = current.next;
    }
      deleteFirst(value);
  }
  public void deleteAt(int index){
    if(last == null || index > getLength()-1){
      return;
    }
    if(last.next == last){
      last = null;
      return;
    }
    DoublyCircNode current = last;
    int i = 0;
    while(current.next != last && i < index){
      current = current.next;
      i++;
    }
    if(current == current.next.next){
      last.next = last;
      last.prev = last;
    }
    else if(current.next == last){
      current.next = current.next.next;
      current.next.prev = current;
      last = current;
    }else{
      current.next = current.next.next;
      current.next.prev = current;
    }
  }
    
  public ArrayList<Integer> getValues(){
    ArrayList<Integer> rlist = new ArrayList<Integer>();
    DoublyCircNode current = last.next;
    while(current != last){
      rlist.add(current.value);
      current = current.next;
    }
    rlist.add(last.value);
    return rlist;
  }
  
  public void reverse(){
    if(last == null){
      return;
    }
    DoublyCircNode current = last.next;
    DoublyCircNode previous = last;
    DoublyCircNode next = last.next;
    
    while(current != last){
      next = current.next;
      current.next = previous;
      current.prev = next;
      previous = current;
      current = next;
    }
      next = current.next;
      current.next = previous;
      current.prev = next;
      last = next;  
  }

public DoublyCircNode reverseRecursive(DoublyCircNode previous, DoublyCircNode current, DoublyCircNode stop){
  DoublyCircNode next;
  if(current == stop){
     next = current.next;
    current.next = previous;
    current.prev = next;
    last = next;
    return last;
  }
  next = current.next;
  current.next = previous;
  current.prev = next;
  
  return(reverseRecursive(current, next, stop));  
}
public void setLast(DoublyCircNode value){
last = value;
}
public DoublyCircNode getLast(){
  return last;
}

}
class MyCode {
	public static void main (String[] args) {
	DoublyCircLinkedList l = new DoublyCircLinkedList();
  l.append(1);
  l.append(2);
  l.append(3);
  l.prepend(0);
  l.insert(4,4);
  l.append(4);
  l.deleteFirst(4);
  l.append(3);
  l.prepend(3);
  l.deleteAll(3);
  l.deleteAt(3);
  l.append(4);
  l.append(5);
  
  System.out.println(l.getValues());
  l.reverse();
  System.out.println(l.getValues());
  l.reverseRecursive(l.getLast(), l.getLast().next, l.getLast());
  System.out.println(l.getValues());
	}
}

File: Linked Lists/doubly_circularly_linked_list.py
class DoublyCircNode(object):
    def __init__ (self, d, n = None, p = None):
        self.data = d
        self.next = n
        self.prev = p
    def get_data(self):
        return self.data

    def set_next(self, new_next):
        self.next = new_next
    def set_prev(self, new_prev):
        self.prev = new_prev

class DoublyCircLinkedList(object):
    def __init__(self, last = None):
        self.last = last
        
    def append(self, value):
        newNode = DoublyCircNode(value)
        if(self.last == None):
            self.last = newNode
            self.last.set_next(newNode)
            self.last.set_prev(newNode)
            return
        newNode.set_prev(self.last)
        newNode.set_next(self.last.next)
        self.last.next.set_prev(newNode)
        self.last.set_next(newNode)
        self.last = newNode
        
    def prepend(self, value):
        newNode = DoublyCircNode(value)
        if(self.last == None):
            self.append(value)
            return
        newNode.set_prev(self.last)
        newNode.set_next(self.last.next)
        self.last.next.set_prev(newNode)
        self.last.set_next(newNode)
        return
    def getLength(self):
        if(self.last != None):
            current = self.last
            i = 1
            while(current.next != self.last):
                i+=1
                current = current.next
            return i
        else:
            return 0
        
    def insert(self, value, index):
        if(index == 0):
            prepend(value)
            return
        elif(index > self.getLength() or self.last == None):
            append(value)
            return
        else:
            current = self.last
            i = 1
            while(current.next != self.last and i <index):
                i+=1
                current = current.next
            newNode = DoublyCircNode(value)
            if(current.next == self.last):
                newNode.set_prev(current)
                newNode.set_next(current.next)
                current.next.set_prev(newNode)
                current.set_next(newNode)
            else:
                newNode.set_prev(current)
                newNode.set_next(current.next)
                current.next.set_prev(newNode)
                current.set_next(newNode)
                
    def deleteFirst(self, value):
        current = self.last
        if(self.last == None):
            return
        elif(self.last.next == self.last and self.last.get_data() != value):
            self.last = None
            return
        while(current.next != self.last and current.next.get_data() != value):
            current = current.next
        if(current.next == self.last and self.last.get_data() == value):
            self.last.prev.set_next(self.last.next)
            self.last.next.set_prev(self.last.prev)
            self.last = self.last.prev
            return
        elif(current.next == self.last and self.last.get_data() != value):
            return
        else:
            current.set_next(current.next.next)
            current.next.set_prev(current)
            return
    
    def deleteAll(self, value):
        if(self.last == None):
            return
        current = self.last.next
        while(current != self.last):
            if(current.get_data() == value):
                self.deleteFirst(value)
            current = current.next
        self.deleteFirst(value)
            
    def deleteAt(self, index):
        if(self.last == None or index > self.getLength()-1):
            return
        if(self.last.next == self.last):
            self.last = None
            return
        current = self.last
        i = 0
        while(current.next != self.last and i < index):
            current = current.next
            i+=1
        if(current == current.next.next):
            self.last.set_next(self.last)
            self.last.set_prev(self.last)
        elif(current.next == self.last):
            current.set_next(current.next.next)
            current.next.set_prev(current)
            self.last = current
        else:
            current.set_next(current.next.next)
            current.next.set_prev(current)
            
    def reverse(self):
        if(self.last == None):
            return
        current = self.last.next
        previous = self.last
        n = self.last.next
        while(current != self.last):
            n = current.next
            current.set_next(previous)
            current.set_prev(n)
            previous = current
            current = n
        n = current.next
        current.set_next(previous)
        current.set_prev(n)
        self.last = n
        
        
    def reverseRecursive(self, previous, current, stop):
        if(current == stop):
            n = current.next
            current.set_next(previous)
            current.set_prev(next)
            self.last = n
            return self.last
        n = current.next
        current.set_next(previous)
        current.set_prev(next)
        
        return (self.reverseRecursive(current, n, stop))
        
    def get_last(self):
        return self.last
    def set_last(self, value):
        self.last = DoublyCircNode(value)
    def getValues(self):
        arr = []
        current = self.last.next
        while(current != self.last):
            arr.append(current.get_data())
            current = current.next
        arr.append(self.last.get_data())
        return arr
def main():
        l = DoublyCircLinkedList()
        l.append(3)
        l.append(2)
        l.append(1)
        l.append(0)
        l.prepend(4)
        l.insert(3,3)
        l.deleteFirst(3)
        l.prepend(2)
        l.append(2)
        l.deleteAll(2)
        l.deleteAt(2)
        
        print(l.getValues())
        l.reverse()
        print(l.getValues())
        l.reverseRecursive(l.get_last(), l.get_last().next, l.get_last())
        print(l.getValues())
        
        
if __name__ == "__main__":
    main()
